### PR TITLE
Issue #493 -  GTID gaps in binlog

### PR DESCRIPTION
**Pull request by kirjaamo**
A new troubleshooting article added for GTID gaps in binlog.

### CODE REVIEW

File: galeracluster/source/kb/gaps-in-gtid-sequence-numbers.rst [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+   SET @@SESSION.GTID_NEXT= 'a9264492-7be5-11ef-b353-ca382343a2f9:2'
+   use `test`; create table t2 (i int primary key, j int, foreign key(j) references t1(i)) /* xid=4 */
+   SET @@SESSION.GTID_NEXT= 'a9264492-7be5-11ef-b353-ca382343a2f9:4'
+   use `test`; create table t3 (i int primary key, j int, foreign key(j) references t1(i)) /* xid=6 */
+
+Here, GTID seqno 3 is allocated for the ``create table`` statement, but the related GTID event is missing from the binlog file.
+
+The excerpt from the binlog suggests that the ``gtid_next`` value is manually set before every executed transaction. For a failed transaction, this GTID value is not binlogged, although it is counted as used.
+
+In other words, for GTID values manually set with ``gtid_next``, there is a possibility of creating a gap in the binlog. This is allowed to happen, as the replication subsystem is based on the ``gtid_executed`` variable, when determining which events should be replicated between the primary and the replica, and not on the binlog contents. Binlog solely becomes a storage for all GTIDs executed, to transfer them over to the replica.
```
Comment (denis-protivensky) [inline galeracluster/source/kb/gaps-in-gtid-sequence-numbers.rst:94]: I think it's better to rephrase this whole block with something like:

"Galera cluster replication is based on the assumption that GTID numbers don't contain gaps in the binlog. In such a case the behavior is undefined, thus it should be avoided".

File: galeracluster/source/kb/gaps-in-gtid-sequence-numbers.rst [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+.. code-block:: mysql
+
+   # Record some GTID
+   SET @@SESSION.GTID_NEXT= 'a9264492-7be5-11ef-b353-ca382343a2f9:1';
+   CREATE TABLE t1 (i INT PRIMARY KEY, j INT) ENGINE=InnoDB;
+   
+   # Now miss a GTID value by failing a transaction
+   SET @@SESSION.GTID_NEXT= 'a9264492-7be5-11ef-b353-ca382343a2f9:2';
+   --error 0, 1824
```
Comment (denis-protivensky) [inline galeracluster/source/kb/gaps-in-gtid-sequence-numbers.rst:109]: Technically, this is an MTR runner code, not something that the mysql shell accepts AFAIK. So at least this `--error 0, 1824` should be removed maybe?

File: galeracluster/source/kb/gaps-in-gtid-sequence-numbers.rst [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+.. rst-class:: section-heading
+.. rubric:: Scenario
+
+When a failing DDL statement with a GTID seqno is replicated by TOI, the failed DDL is not logged in the binary log and, therefore, the binlog will contain GTID event gaps.
+
+Below is an example of a failing DDL statement:
+
+.. code-block:: mysql
+
+   create table t_fail (i int primary key, j int, foreign key(j) references non_existing(i))
```
Comment (denis-protivensky) [inline galeracluster/source/kb/gaps-in-gtid-sequence-numbers.rst:77]: You need to mention that
`SET @@SESSION.GTID_NEXT= 'a9264492-7be5-11ef-b353-ca382343a2f9:3';`
command is issued before this failing create table.

And also add a terminating semicolon after create table statement.

File: galeracluster/source/kb/gaps-in-gtid-sequence-numbers.rst [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+.. rst-class:: article-stats
+
+   Length: 421 words; Published: October 22, 2024; Updated: October 22, 2024; Category: State Transfers; Type: Troubleshooting
+
+GTID gaps may occur if ``gtid_next`` is set to a specific value, and a transaction fails.
+
+.. rst-class:: section-heading
+.. rubric:: Scenario
+
+When a failing DDL statement with a GTID seqno is replicated by TOI, the failed DDL is not logged in the binary log and, therefore, the binlog will contain GTID event gaps.
```
Comment (denis-protivensky) [inline galeracluster/source/kb/gaps-in-gtid-sequence-numbers.rst:71]: You'd better start like that:
"When a failing DDL statement with explicit GTID value set with `gtid_next` is replicated ..."

Note: technically, any DDL or DML is affected with explicit `gtid_next` setting, but as an example scenario the current version looks OK.

**Review by denis-protivensky — CHANGES_REQUESTED** [review]

File: galeracluster/source/kb/gaps-in-gtid-sequence-numbers.rst (hunk 1 above)
Comment (kirjaamo) [reply galeracluster/source/kb/gaps-in-gtid-sequence-numbers.rst]: Fixed

File: galeracluster/source/kb/gaps-in-gtid-sequence-numbers.rst (hunk 2 above)
Comment (kirjaamo) [reply galeracluster/source/kb/gaps-in-gtid-sequence-numbers.rst]: Fixed

File: galeracluster/source/kb/gaps-in-gtid-sequence-numbers.rst (hunk 3 above)
Comment (kirjaamo) [reply galeracluster/source/kb/gaps-in-gtid-sequence-numbers.rst]: Fixed

**Comment by kirjaamo** [comment]
@byte Denis approved. Can you merge?

**Comment by kirjaamo** [comment]
@denis-protivensky Can you confirm that these are ok, and if they are, please approve and merge.

**Review by byte — APPROVED** [review]
ok to push

**Pull request merged**